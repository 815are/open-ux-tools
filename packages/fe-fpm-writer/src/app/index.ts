import { create as createStorage } from 'mem-fs';
import type { Editor } from 'mem-fs-editor';
import { create } from 'mem-fs-editor';
import { join } from 'path';
import { satisfies } from 'semver';
import type { SAPJSONSchemaForWebApplicationManifestFile as Manifest } from '../common/manifest';

export interface FPMConfig {
    fcl?: boolean;
}

/**
 * Enable the flexible programming model for an application.
 *
 * @param {string} basePath - the base path
 * @param {CustomTableColumn} config - the custom column configuration
 * @param {Editor} [fs] - the mem-fs editor instance
 * @returns {Promise<Editor>} the updated mem-fs editor instance
 */
export function enableFPM(basePath: string, config: FPMConfig = {}, fs?: Editor): Editor {
    if (!fs) {
        fs = create(createStorage());
    }

    const manifestPath = join(basePath, 'webapp/manifest.json');
    const manifest = fs.readJSON(manifestPath) as any as Manifest;

    // add FE libs is not yet added
    if (!manifest['sap.ui5']?.dependencies?.libs?.['sap.fe.templates']) {
        fs.extendJSON(manifestPath, {
            'sap.ui5': {
                dependencies: {
                    libs: {
                        'sap.fe.templates': {}
                    }
                }
            }
        });
    }

    // if a minUI5Version is set and it is smaller than the minimum required, increase it
    if (
        manifest['sap.ui5']?.dependencies.minUI5Version &&
        !satisfies(manifest['sap.ui5']?.dependencies.minUI5Version, '>=1.86')
    ) {
        fs.extendJSON(manifestPath, {
            'sap.ui5': {
                dependencies: {
                    minUI5Version: '1.86.0'
                }
            }
        });
    }
    // enable FCL if requested
    if (config.fcl) {
        fs.extendJSON(manifestPath, {
            'sap.ui5': {
                rootView: {
                    viewName: 'sap.fe.templates.RootContainer.view.Fcl',
                    type: 'XML',
                    async: true,
                    id: 'appRootView'
                },
                routing: {
                    config: {
                        routerClass: 'sap.f.routing.Router'
                    }
                }
            }
        });
    }

    // replace Component.js
    const componentTemplate = join(__dirname, '../../templates/app/Component.js');
    fs.copyTpl(componentTemplate, join(basePath, 'webapp/Component.js'), manifest['sap.app']);

    return fs;
}
