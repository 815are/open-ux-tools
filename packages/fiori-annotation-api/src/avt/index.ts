export { convertAnnotationFile } from './annotations';
export { convertMetadataToAvtSchema } from './metadata';
