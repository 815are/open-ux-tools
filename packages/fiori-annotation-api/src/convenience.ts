import type {
    RawSchema,
    RawEntityContainer,
    RawEntitySet,
    RawEntityType,
    RawComplexType,
    RawAction,
    RawProperty,
    RawV4NavigationProperty,
    AnnotationList,
    ActionParameter
} from '@sap-ux/vocabularies-types';
import type { MetadataService, MetadataElement, Path } from '@sap-ux/odata-metadata';

/**
 * convert metadata to AVT Schema representation
 * @param metadataService
 */
export function convertMetaDataToAvtSchema(metadataService: MetadataService): RawSchema {
    const rootLevelElement = metadataService.getRootMetadataElements();
    // Schema representation includes entityTypes, entitySets, associations (not supported here), a single entityContainer and actions
    // collect these objects based on the (multiple) EdmTargetKinds of each metadata element
    const entityTypes: RawEntityType[] = [];
    const complexTypes: RawComplexType[] = [];
    const entitySets: RawEntitySet[] = [];
    let entityContainer: RawEntityContainer = {
        _type: 'EntityContainer',
        fullyQualifiedName: ''
    };
    const actions: RawAction[] = [];
    rootLevelElement.forEach((element: MetadataElement, elementKey: Path) => {
        const targetKinds = metadataService.getEdmTargetKinds(element.path);
        if (targetKinds.includes('EntityContainer')) {
            entityContainer = {
                _type: 'EntityContainer',
                name: element.name,
                fullyQualifiedName: elementKey
            };
            // entity sets as children of entity container
            for (const subElement of element.content ?? []) {
                const subElementTargetKinds = metadataService.getEdmTargetKinds(subElement.path);
                if (subElementTargetKinds.includes('EntitySet')) {
                    // only EntitySets are considered in AVT Schema representation (no Singleton, Action/FunctionImport etc...)
                    const entitySet: RawEntitySet = {
                        _type: 'EntitySet',
                        name: subElement.name,
                        entityTypeName: subElement.structuredType ?? '', // TODO: check if this causes issues
                        navigationPropertyBinding: {},
                        fullyQualifiedName: subElement.path
                    };
                    entitySets.push(entitySet);
                }
                if (metadataService.ODataVersion === '2.0' && subElementTargetKinds.includes('FunctionImport')) {
                    // remark: AVT fills it differently in parseFunctionImport() - but then annotations are not picked up in convertTypes()
                    const action: RawAction = {
                        _type: 'Action',
                        name: subElement.name,
                        fullyQualifiedName: subElement.path,
                        parameters: [],
                        isBound: false, // always set to false in AVT!
                        sourceType: '', // not available here TODO: check if this causes issues
                        returnType: '', // TODO: check if this causes issues
                        isFunction: false
                    };
                    for (const parameterMdElement of subElement.content ?? []) {
                        if (parameterMdElement) {
                            const parameterTargetKinds = metadataService.getEdmTargetKinds(parameterMdElement.path);
                            if (parameterTargetKinds.includes('Parameter')) {
                                // Set action.sourceType to the type from the parameter '_it'
                                if (parameterMdElement.name === '_it') {
                                    action.sourceType = parameterMdElement.structuredType ?? ''; // TODO: check if this causes issues
                                }
                                const parameter: ActionParameter = {
                                    _type: 'ActionParameter',
                                    isEntitySet: false,
                                    fullyQualifiedName: parameterMdElement.path,
                                    annotations: {},
                                    name: parameterMdElement.name,
                                    type: parameterMdElement.structuredType ?? parameterMdElement.edmPrimitiveType ?? '' // TODO: check if this causes issues
                                };
                                action.parameters.push(parameter);
                            } else if (parameterTargetKinds.includes('ReturnType')) {
                                action.returnType =
                                    parameterMdElement.structuredType ?? parameterMdElement.edmPrimitiveType ?? ''; // TODO: check if this causes issues
                            }
                        }
                    }
                    actions.push(action);
                }
            }
        }
        if (targetKinds.includes('EntityType')) {
            const entityProperties: RawProperty[] = [];
            const entityNavProps: RawV4NavigationProperty[] = [];
            const keys: RawProperty[] = [];
            for (const subElement of element.content ?? []) {
                const subElementTargetKinds = metadataService.getEdmTargetKinds(subElement.path);
                if (subElementTargetKinds.includes('Property')) {
                    const property: RawProperty = {
                        _type: 'Property',
                        name: subElement.name,
                        type: subElement.edmPrimitiveType ?? subElement.structuredType ?? '', // TODO: check if this causes issues
                        fullyQualifiedName: subElement.path
                    };
                    entityProperties.push(property);
                    if (element.keys && element.keys.includes(subElement.name)) {
                        keys.push(property);
                    }
                }
                if (subElementTargetKinds.includes('NavigationProperty')) {
                    const navProp: RawV4NavigationProperty = {
                        _type: 'NavigationProperty',
                        name: subElement.name,
                        fullyQualifiedName: subElement.path,
                        targetTypeName: subElement.structuredType ?? '', // TODO: check if this causes issues
                        isCollection: false,
                        containsTarget: false,
                        partner: '', // TODO: check if this causes issues
                        referentialConstraint: []
                    };
                    entityNavProps.push(navProp);
                }
            }
            const entityType: RawEntityType = {
                _type: 'EntityType',
                name: element.name,
                fullyQualifiedName: elementKey,
                entityProperties: entityProperties,
                keys: keys,
                navigationProperties: entityNavProps,
                actions: {}
            };
            entityTypes.push(entityType);
        }
        if (targetKinds.includes('ComplexType')) {
            const complexTypeProperties: RawProperty[] = [];
            const complexTypeNavProperties: RawV4NavigationProperty[] = [];
            for (const subElement of element.content ?? []) {
                const subElementTargetKinds = metadataService.getEdmTargetKinds(subElement.path);
                if (subElementTargetKinds.includes('Property')) {
                    const property: RawProperty = {
                        _type: 'Property',
                        name: subElement.name,
                        type: subElement.edmPrimitiveType ?? subElement.structuredType ?? '', // TODO: check if this causes issues
                        fullyQualifiedName: subElement.path
                    };
                    complexTypeProperties.push(property);
                }
                if (subElementTargetKinds.includes('NavigationProperty')) {
                    const navProp: RawV4NavigationProperty = {
                        _type: 'NavigationProperty',
                        name: subElement.name,
                        fullyQualifiedName: subElement.path,
                        targetTypeName: subElement.structuredType ?? '', // TODO: check if this causes issues
                        isCollection: false,
                        containsTarget: false,
                        partner: '', // TODO: check if this causes issues
                        referentialConstraint: []
                    };
                    complexTypeNavProperties.push(navProp);
                }
            }
            const complexType: RawComplexType = {
                _type: 'ComplexType',
                name: element.name,
                fullyQualifiedName: elementKey,
                properties: complexTypeProperties,
                navigationProperties: complexTypeNavProperties
            };
            complexTypes.push(complexType);
        }
        if (targetKinds.includes('EntitySet')) {
            // for CDS no entity container is present: target kind EntitySet can appear at root level
            const entitySet: RawEntitySet = {
                _type: 'EntitySet',
                name: element.name,
                entityTypeName: element.structuredType || (targetKinds.includes('EntityType') ? element.name : ''),
                navigationPropertyBinding: {},
                fullyQualifiedName: element.path
            };
            entitySets.push(entitySet);
        }
        if (targetKinds.includes('Action') || targetKinds.includes('Function')) {
            // see AVT parseActions()
            const action: RawAction = {
                _type: 'Action',
                name: element.name,
                fullyQualifiedName: element.path,
                parameters: [],
                isBound: false, // TODO: check if this causes issues
                sourceType: '', // not available here // TODO: check if this causes issues
                returnType: '', // TODO: check if this causes issues
                isFunction: false
            };
            for (const subElement of element.content ?? []) {
                const subElementTargetKinds = metadataService.getEdmTargetKinds(subElement.path);
                if (subElementTargetKinds.includes('Parameter')) {
                    const parameter = {
                        _type: 'ActionParameter',
                        isEntitySet: false,
                        fullyQualifiedName: subElement.path,
                        type: subElement.structuredType || subElement.edmPrimitiveType
                    };
                    action.parameters.push(parameter as any);
                } else if (subElementTargetKinds.includes('ReturnType')) {
                    action.returnType = subElement.structuredType ?? subElement.edmPrimitiveType ?? ''; // TODO: check if this causes issues
                }
            }
            actions.push(action);
        }
    });

    const [namespace] = [...metadataService.getNamespaces()];
    if (entityContainer.fullyQualifiedName && namespace) {
        // for CDS no entity container is present: build dummy entity container
        const entityContainerName = 'dummyEntityContainer';
        entityContainer = {
            _type: 'EntityContainer',
            name: entityContainerName,
            fullyQualifiedName: namespace ? namespace + '.' + entityContainerName : entityContainerName
        };
    }

    const annotations: Record<string, AnnotationList[]> = {};
    return {
        namespace,
        annotations,
        entitySets,
        complexTypes,
        entityContainer,
        actions,
        entityTypes,
        associations: [],
        singletons: [],
        associationSets: [],
        typeDefinitions: []
    };
}
