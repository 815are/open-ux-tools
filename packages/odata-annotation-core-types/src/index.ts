export * from './annotation-file';
export * from './base';
export * from './edm';
export * from './edmx';

export { Location, Range, Position, Diagnostic } from './language-server';
