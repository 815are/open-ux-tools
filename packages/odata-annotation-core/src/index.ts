export * from '@sap-ux/odata-annotation-core-types';
export {
    ParsedName,
    ParsedActionFunctionSignature,
    ParsedCollectionIdentifier,
    ParsedIdentifier,
    parseIdentifier,
    toFullyQualifiedName
} from './names';

export {
    NavigationPropertyAnnotationSegment,
    ParsedPath,
    ParsedPathSegment,
    TermCastSegment,
    parsePath,
    toFullyQualifiedPath
} from './paths';
