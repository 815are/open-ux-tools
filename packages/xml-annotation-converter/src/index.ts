export { convertDocument, convertMetadataDocument } from './parser';
