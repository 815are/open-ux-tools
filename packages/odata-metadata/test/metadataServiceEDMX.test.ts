import { MetadataService } from '../src/metadataService';
import type { MetadataElement, ODataVersionType } from '../src/types';

function createTestService() {
    const ODataVersion: ODataVersionType = '2.0';
    const metadataService = new MetadataService({ ODataVersion });

    const metadataFileUri = 'file://metadata.xml';
    const metadataElements: MetadataElement[] = [
        {
            location: {
                uri: metadataFileUri,
                range: { start: { line: 1, character: 1 }, end: { line: 5, character: 5 } }
            },
            isAnnotatable: true,
            kind: 'EntityType',
            path: '/com.sap.gateway.default.iwbep.tea_busi.v0001.Department',
            name: 'com.sap.gateway.default.iwbep.tea_busi.v0001.Department',
            isCollectionValued: false,
            isComplexType: false,
            isEntityType: true,
            content: [
                {
                    path: '/com.sap.gateway.default.iwbep.tea_busi.v0001.Department/Sector',
                    content: [],
                    isAnnotatable: true,
                    kind: 'Property',
                    name: 'Sector',
                    isCollectionValued: false,
                    isComplexType: false,
                    isEntityType: false,
                    edmPrimitiveType: 'Edm.String'
                },
                {
                    path: '/com.sap.gateway.default.iwbep.tea_busi.v0001.Department/ID',
                    content: [],
                    isAnnotatable: true,
                    kind: 'Property',
                    name: 'ID',
                    isCollectionValued: false,
                    isComplexType: false,
                    isEntityType: false,
                    edmPrimitiveType: 'Edm.String'
                },
                {
                    path: '/com.sap.gateway.default.iwbep.tea_busi.v0001.Department/Name',
                    isAnnotatable: true,
                    kind: 'Property',
                    name: 'Name',
                    isCollectionValued: false,
                    isComplexType: false,
                    isEntityType: false,
                    edmPrimitiveType: 'Edm.String'
                },
                {
                    path: '/com.sap.gateway.default.iwbep.tea_busi.v0001.Department/MemberCount',
                    isAnnotatable: true,
                    kind: 'Property',
                    name: 'MemberCount',
                    isCollectionValued: false,
                    isComplexType: false,
                    isEntityType: false,
                    edmPrimitiveType: 'Edm.Int32'
                },
                {
                    path: '/com.sap.gateway.default.iwbep.tea_busi.v0001.Department/ManagerID',
                    isAnnotatable: true,
                    kind: 'Property',
                    name: 'ManagerID',
                    isCollectionValued: false,
                    isComplexType: false,
                    isEntityType: false,
                    edmPrimitiveType: 'Edm.String'
                },
                {
                    path: '/com.sap.gateway.default.iwbep.tea_busi.v0001.Department/Categories',
                    isAnnotatable: true,
                    kind: 'Property',
                    name: 'Categories',
                    isCollectionValued: true,
                    isComplexType: false,
                    isEntityType: false,
                    edmPrimitiveType: 'Edm.String'
                }
            ]
        },
        // function overloads (function 'com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment' with different signature)
        {
            isAnnotatable: true,
            kind: 'Function',
            path: '/com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.String)',
            name: 'com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.String)',
            isCollectionValued: false,
            isComplexType: false,
            isEntityType: false,
            content: [
                {
                    path: '/com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.String)/DepartmentName',
                    isAnnotatable: true,
                    kind: 'Parameter',
                    name: 'DepartmentName',
                    isCollectionValued: false,
                    isComplexType: false,
                    isEntityType: false,
                    edmPrimitiveType: 'Edm.String'
                }
            ]
        },
        {
            isAnnotatable: true,
            kind: 'Function',
            path: '/com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.Int16)',
            name: 'com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.Int16)',
            isCollectionValued: false,
            isComplexType: false,
            isEntityType: false,
            content: [
                {
                    path: '/com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.Int16)/DepartmentID',
                    isAnnotatable: true,
                    kind: 'Parameter',
                    name: 'DepartmentID',
                    isCollectionValued: false,
                    isComplexType: false,
                    isEntityType: false,
                    edmPrimitiveType: 'Edm.Int16'
                }
            ]
        }
    ];
    metadataService.import(metadataElements);
    return metadataService;
}

describe('MetadataService for XML', () => {
    test('import()', () => {
        // Prepare
        const callbackParameters: { element: MetadataElement }[] = [];
        const localMetadataService = createTestService();
        localMetadataService.visitMetadataElements(function (element: MetadataElement) {
            callbackParameters.push({ element });
        });
        // Expect
        expect(callbackParameters).toMatchSnapshot();
    });

    test('getNamespaces()', () => {
        // Expect
        expect(createTestService().getNamespaces()).toMatchInlineSnapshot(`
            Set {
              "com.sap.gateway.default.iwbep.tea_busi.v0001",
            }
        `);
    });

    test('getRootMetadataElements()', () => {
        // Expect
        expect(createTestService().getRootMetadataElements()).toMatchInlineSnapshot(`
            Map {
              "com.sap.gateway.default.iwbep.tea_busi.v0001.Department" => Object {
                "fileUri": "someAbsoluteFileUri",
                "isAnnotatable": true,
                "isCollectionValued": false,
                "isComplexType": false,
                "isEntityType": true,
                "kind": "EntityType",
                "name": "com.sap.gateway.default.iwbep.tea_busi.v0001.Department",
                "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.Department",
                "range": Object {
                  "end": Object {
                    "character": 5,
                    "line": 5,
                  },
                  "start": Object {
                    "character": 1,
                    "line": 1,
                  },
                },
                "subElementNames": Set {
                  "Sector",
                  "ID",
                  "Name",
                  "MemberCount",
                  "ManagerID",
                  "Categories",
                },
              },
              "com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.String)" => Object {
                "fileUri": "someAbsoluteFileUri",
                "isAnnotatable": true,
                "isCollectionValued": false,
                "isComplexType": false,
                "isEntityType": false,
                "kind": "Function",
                "name": "com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.String)",
                "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.String)",
                "range": Object {
                  "end": Object {
                    "character": 1,
                    "line": 8,
                  },
                  "start": Object {
                    "character": 1,
                    "line": 6,
                  },
                },
                "subElementNames": Set {
                  "DepartmentName",
                },
              },
              "com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.Int16)" => Object {
                "fileUri": "someAbsoluteFileUri",
                "isAnnotatable": true,
                "isCollectionValued": false,
                "isComplexType": false,
                "isEntityType": false,
                "kind": "Function",
                "name": "com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.Int16)",
                "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.Int16)",
                "range": Object {
                  "end": Object {
                    "character": 1,
                    "line": 11,
                  },
                  "start": Object {
                    "character": 1,
                    "line": 9,
                  },
                },
                "subElementNames": Set {
                  "DepartmentID",
                },
              },
            }
        `);
    });

    test('getMetadataElement()', () => {
        // Expect
        expect(createTestService().getMetadataElement('com.sap.gateway.default.iwbep.tea_busi.v0001.Department'))
            .toMatchInlineSnapshot(`
            Object {
              "fileUri": "someAbsoluteFileUri",
              "isAnnotatable": true,
              "isCollectionValued": false,
              "isComplexType": false,
              "isEntityType": true,
              "kind": "EntityType",
              "name": "com.sap.gateway.default.iwbep.tea_busi.v0001.Department",
              "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.Department",
              "range": Object {
                "end": Object {
                  "character": 5,
                  "line": 5,
                },
                "start": Object {
                  "character": 1,
                  "line": 1,
                },
              },
              "subElementNames": Set {
                "Sector",
                "ID",
                "Name",
                "MemberCount",
                "ManagerID",
                "Categories",
              },
            }
        `);
    });

    test('getMetadataElement() for property', () => {
        // Expect
        expect(
            createTestService().getMetadataElement('com.sap.gateway.default.iwbep.tea_busi.v0001.Department/Categories')
        ).toMatchInlineSnapshot(`
            Object {
              "edmPrimitiveType": "Edm.String",
              "isAnnotatable": true,
              "isCollectionValued": true,
              "isComplexType": false,
              "isEntityType": false,
              "kind": "Property",
              "name": "Categories",
              "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.Department/Categories",
              "range": undefined,
              "subElementNames": Set {},
            }
        `);
    });

    test('getMetadataElement() for function', () => {
        // Expect
        expect(createTestService().getMetadataElement('com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment'))
            .toMatchInlineSnapshot(`
            Object {
              "fileUri": "someAbsoluteFileUri",
              "isAnnotatable": true,
              "isCollectionValued": false,
              "isComplexType": false,
              "isEntityType": false,
              "kind": "Function",
              "name": "com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.String)",
              "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.String)",
              "range": Object {
                "end": Object {
                  "character": 1,
                  "line": 8,
                },
                "start": Object {
                  "character": 1,
                  "line": 6,
                },
              },
              "subElementNames": Set {
                "DepartmentName",
              },
            }
        `);
    });

    test('getMetadataElement() for parameter of function', () => {
        // Expect
        expect(
            createTestService().getMetadataElement(
                'com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment/DepartmentID'
            )
        ).toMatchInlineSnapshot(`
            Object {
              "edmPrimitiveType": "Edm.Int16",
              "isAnnotatable": true,
              "isCollectionValued": false,
              "isComplexType": false,
              "isEntityType": false,
              "kind": "Parameter",
              "name": "DepartmentID",
              "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.Int16)/DepartmentID",
              "range": undefined,
              "subElementNames": Set {},
            }
        `);
    });

    test('getMetadataElement() for specific function overload', () => {
        // Expect
        expect(
            createTestService().getMetadataElement(
                'com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.Int16)'
            )
        ).toMatchInlineSnapshot(`
            Object {
              "fileUri": "someAbsoluteFileUri",
              "isAnnotatable": true,
              "isCollectionValued": false,
              "isComplexType": false,
              "isEntityType": false,
              "kind": "Function",
              "name": "com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.Int16)",
              "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.Int16)",
              "range": Object {
                "end": Object {
                  "character": 1,
                  "line": 11,
                },
                "start": Object {
                  "character": 1,
                  "line": 9,
                },
              },
              "subElementNames": Set {
                "DepartmentID",
              },
            }
        `);
    });

    test('getMetadataElement() for parameter of specific function overload', () => {
        // Expect
        expect(
            createTestService().getMetadataElement(
                'com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.Int16)/DepartmentID'
            )
        ).toMatchInlineSnapshot(`
            Object {
              "edmPrimitiveType": "Edm.Int16",
              "isAnnotatable": true,
              "isCollectionValued": false,
              "isComplexType": false,
              "isEntityType": false,
              "kind": "Parameter",
              "name": "DepartmentID",
              "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.Int16)/DepartmentID",
              "range": undefined,
              "subElementNames": Set {},
            }
        `);
    });

    test('getEdmTargetKinds()', () => {
        // Expect
        expect(createTestService().getEdmTargetKinds('com.sap.gateway.default.iwbep.tea_busi.v0001.Department'))
            .toMatchInlineSnapshot(`
            Array [
              "EntityType",
            ]
        `);
    });

    test('getEdmTargetKinds() for collection valued property', () => {
        // Expect
        expect(
            createTestService().getEdmTargetKinds('com.sap.gateway.default.iwbep.tea_busi.v0001.Department/Categories')
        ).toMatchInlineSnapshot(`
                      Array [
                        "Property",
                        "Collection",
                      ]
              `);
    });

    test('getMetadataElementLocations() for EntityType', () => {
        expect(
            createTestService().getMetadataElementLocations('com.sap.gateway.default.iwbep.tea_busi.v0001.Department')
        ).toMatchInlineSnapshot(`
            Object {
              "range": Object {
                "end": Object {
                  "character": 5,
                  "line": 5,
                },
                "start": Object {
                  "character": 1,
                  "line": 1,
                },
              },
              "uri": "someAbsoluteFileUri",
            }
        `);
    });

    test('getMetadataElementLocations() for Function name', () => {
        expect(
            createTestService().getMetadataElementLocations(
                'com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment'
            )
        ).toMatchInlineSnapshot(`
            Array [
              Object {
                "range": Object {
                  "end": Object {
                    "character": 1,
                    "line": 8,
                  },
                  "start": Object {
                    "character": 1,
                    "line": 6,
                  },
                },
                "uri": "someAbsoluteFileUri",
              },
              Object {
                "range": Object {
                  "end": Object {
                    "character": 1,
                    "line": 11,
                  },
                  "start": Object {
                    "character": 1,
                    "line": 9,
                  },
                },
                "uri": "someAbsoluteFileUri",
              },
            ]
        `);
    });

    test('getMetadataElementLocations() for Function overload', () => {
        expect(
            createTestService().getMetadataElementLocations(
                'com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.String)'
            )
        ).toMatchInlineSnapshot(`
            Object {
              "range": Object {
                "end": Object {
                  "character": 1,
                  "line": 8,
                },
                "start": Object {
                  "character": 1,
                  "line": 6,
                },
              },
              "uri": "someAbsoluteFileUri",
            }
        `);
    });

    test('getEdmTargetKinds() for function import V2', () => {
        // Arrange
        const metadataService = createTestService();
        const metadataElement = {
            isAnnotatable: true,
            kind: 'FunctionImport',
            name: 'ProjectActivation',
            isCollectionValued: false,
            isComplexType: false,
            isEntityType: false,
            path: 'PROJECT_SERVICE.PROJECT_SERVICE_Entities/ProjectActivation',
            subElementNames: new Set([''])
        };
        jest.spyOn(metadataService, 'getMetadataElement').mockReturnValue(metadataElement);
        const path = 'PROJECT_SERVICE.PROJECT_SERVICE_Entities/ProjectActivation';

        // Act
        const result = metadataService.getEdmTargetKinds(path);

        // Expect
        expect(result).toMatchObject(['FunctionImport', 'Action']);
    });
});
