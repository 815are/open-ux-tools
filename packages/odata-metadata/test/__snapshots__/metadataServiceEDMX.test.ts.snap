// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetadataService for XML import() 1`] = `
Array [
  Object {
    "element": Object {
      "fileUri": "someAbsoluteFileUri",
      "isAnnotatable": true,
      "isCollectionValued": false,
      "isComplexType": false,
      "isEntityType": true,
      "kind": "EntityType",
      "name": "com.sap.gateway.default.iwbep.tea_busi.v0001.Department",
      "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.Department",
      "range": Object {
        "end": Object {
          "character": 5,
          "line": 5,
        },
        "start": Object {
          "character": 1,
          "line": 1,
        },
      },
      "subElementNames": Set {
        "Sector",
        "ID",
        "Name",
        "MemberCount",
        "ManagerID",
        "Categories",
      },
    },
    "fileUri": "testFileUri",
  },
  Object {
    "element": Object {
      "edmPrimitiveType": "Edm.String",
      "isAnnotatable": true,
      "isCollectionValued": false,
      "isComplexType": false,
      "isEntityType": false,
      "kind": "Property",
      "name": "Sector",
      "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.Department/Sector",
      "range": undefined,
      "subElementNames": Set {},
    },
    "fileUri": "testFileUri",
  },
  Object {
    "element": Object {
      "edmPrimitiveType": "Edm.String",
      "isAnnotatable": true,
      "isCollectionValued": false,
      "isComplexType": false,
      "isEntityType": false,
      "kind": "Property",
      "name": "ID",
      "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.Department/ID",
      "range": undefined,
      "subElementNames": Set {},
    },
    "fileUri": "testFileUri",
  },
  Object {
    "element": Object {
      "edmPrimitiveType": "Edm.String",
      "isAnnotatable": true,
      "isCollectionValued": false,
      "isComplexType": false,
      "isEntityType": false,
      "kind": "Property",
      "name": "Name",
      "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.Department/Name",
      "range": undefined,
      "subElementNames": Set {},
    },
    "fileUri": "testFileUri",
  },
  Object {
    "element": Object {
      "edmPrimitiveType": "Edm.Int32",
      "isAnnotatable": true,
      "isCollectionValued": false,
      "isComplexType": false,
      "isEntityType": false,
      "kind": "Property",
      "name": "MemberCount",
      "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.Department/MemberCount",
      "range": undefined,
      "subElementNames": Set {},
    },
    "fileUri": "testFileUri",
  },
  Object {
    "element": Object {
      "edmPrimitiveType": "Edm.String",
      "isAnnotatable": true,
      "isCollectionValued": false,
      "isComplexType": false,
      "isEntityType": false,
      "kind": "Property",
      "name": "ManagerID",
      "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.Department/ManagerID",
      "range": undefined,
      "subElementNames": Set {},
    },
    "fileUri": "testFileUri",
  },
  Object {
    "element": Object {
      "edmPrimitiveType": "Edm.String",
      "isAnnotatable": true,
      "isCollectionValued": true,
      "isComplexType": false,
      "isEntityType": false,
      "kind": "Property",
      "name": "Categories",
      "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.Department/Categories",
      "range": undefined,
      "subElementNames": Set {},
    },
    "fileUri": "testFileUri",
  },
  Object {
    "element": Object {
      "fileUri": "someAbsoluteFileUri",
      "isAnnotatable": true,
      "isCollectionValued": false,
      "isComplexType": false,
      "isEntityType": false,
      "kind": "Function",
      "name": "com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.String)",
      "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.String)",
      "range": Object {
        "end": Object {
          "character": 1,
          "line": 8,
        },
        "start": Object {
          "character": 1,
          "line": 6,
        },
      },
      "subElementNames": Set {
        "DepartmentName",
      },
    },
    "fileUri": "testFileUri",
  },
  Object {
    "element": Object {
      "edmPrimitiveType": "Edm.String",
      "isAnnotatable": true,
      "isCollectionValued": false,
      "isComplexType": false,
      "isEntityType": false,
      "kind": "Parameter",
      "name": "DepartmentName",
      "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.String)/DepartmentName",
      "range": undefined,
      "subElementNames": Set {},
    },
    "fileUri": "testFileUri",
  },
  Object {
    "element": Object {
      "fileUri": "someAbsoluteFileUri",
      "isAnnotatable": true,
      "isCollectionValued": false,
      "isComplexType": false,
      "isEntityType": false,
      "kind": "Function",
      "name": "com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.Int16)",
      "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.Int16)",
      "range": Object {
        "end": Object {
          "character": 1,
          "line": 11,
        },
        "start": Object {
          "character": 1,
          "line": 9,
        },
      },
      "subElementNames": Set {
        "DepartmentID",
      },
    },
    "fileUri": "testFileUri",
  },
  Object {
    "element": Object {
      "edmPrimitiveType": "Edm.Int16",
      "isAnnotatable": true,
      "isCollectionValued": false,
      "isComplexType": false,
      "isEntityType": false,
      "kind": "Parameter",
      "name": "DepartmentID",
      "path": "com.sap.gateway.default.iwbep.tea_busi.v0001.GetDepartment(Edm.Int16)/DepartmentID",
      "range": undefined,
      "subElementNames": Set {},
    },
    "fileUri": "testFileUri",
  },
]
`;
